src/foam/graphics/TreeNode.js

```diff
@@ -0,0 +1,297 @@
+/**
+ * @license
+ * Copyright 2019 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.graphics',
+  name: 'TreeNode',
+  extends: 'foam.graphics.Box',
+
+  requires: [
+    'foam.graphics.Box',
+    'foam.graphics.CView',
+    'foam.graphics.Label',
+    'foam.graphics.Line'
+  ],
+
+  imports: [
+    'formatNode',
+    'graph',
+    'nodeHeight',
+    'nodeWidth',
+    'padding',
+    'parentNode?',
+    'relationship'
+  ],
+  exports: ['as parentNode'],
+
+  properties: [
+    'data',
+    {
+      name: 'outline',
+      expression: function (x, nodeWidth, expanded, childNodes, padding) {
+        var nodeLeftEdgePadded = x - nodeWidth / 2 - padding / 2;
+        var nodeRightEdgePadded = x + nodeWidth / 2 + padding / 2;
+        var rootLevelOutline = [
+          {
+            left: nodeLeftEdgePadded,
+            right: nodeRightEdgePadded
+          }
+        ];
+
+        var champion = [];
+
+        for (let i = 0; i < childNodes.length && expanded; i++) {
+          // get child outline
+          // transform all rows
+          // merge levels into childoutlines
+          var childOutline = childNodes[i].outline.map(o => ({
+            left: o.left + x,
+            right: o.right + x,
+          }));
+
+          for (var j = 0; j < childOutline.length; j++) {
+            champion[j] = champion[j] || {
+              left: Number.MAX_SAFE_INTEGER,
+              right: Number.MIN_SAFE_INTEGER
+            };
+            champion[j].left = Math.min(childOutline[j].left, champion[j].left);
+            champion[j].right = Math.max(childOutline[j].right, champion[j].right);
+          }
+        }
+
+        var totalOutline = rootLevelOutline.concat(champion);
```

Comment:
Probably don't need this variable. I'd probably have just renamed `champion` to `outline` and included the root in it so you simply do a `return outline;` as the very last statement.

Reply:
removed it and just did `return rootLevelOutline.concat(outlineBelowRoot);` as I think its a little bit more clearer to what's happening